src/Compilers/Core/Desktop/ConsistencyChecker.cs

```diff
@@ -0,0 +1,137 @@
+    {
+        private readonly ImmutableArray<string> _prefixWhiteList;
+
+        public ConsistencyChecker(IEnumerable<string> prefixWhiteList)
+        {
+            _prefixWhiteList = ImmutableArray.CreateRange(prefixWhiteList);
+        }
+
+        public ImmutableArray<ConsistencyError> CheckAssemblies(Dictionary<string, Assembly> pathsToAssemblies, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            ArrayBuilder<ConsistencyError> errors = new ArrayBuilder<ConsistencyError>();
+
+            Dictionary<string, Guid> pathsToMvids = new Dictionary<string, Guid>(StringComparer.OrdinalIgnoreCase);
+            HashSet<AssemblyIdentity> identitiesOfAssembliesOnDisk = new HashSet<AssemblyIdentity>();
+            Dictionary<string, ImmutableArray<AssemblyIdentity>> pathsToReferences = new Dictionary<string, ImmutableArray<AssemblyIdentity>>(StringComparer.OrdinalIgnoreCase);
+
+            foreach (var path in pathsToAssemblies.Keys)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    using (var reader = new PEReader(FileUtilities.OpenRead(path)))
+                    {
+                        var metadataReader = reader.GetMetadataReader();
+
+                        pathsToMvids[path] = metadataReader.GetModuleVersionIdOrThrow();
+                        identitiesOfAssembliesOnDisk.Add(metadataReader.ReadAssemblyIdentityOrThrow());
+                        pathsToReferences[path] = metadataReader.GetReferencedAssembliesOrThrow();
+                    }
+                }
+                catch (Exception e)
+                {
+                    errors.Add(new ConsistencyError(ConsistencyErrorKind.UnableToReadFile, path, exception: e));
+                }
+            }
+
+            if (errors.Count > 0)
+            {
+                return errors.ToImmutable();
+            }
+
+            foreach (var path in pathsToAssemblies.Keys)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                Guid fileMvid = pathsToMvids[path];
+                Guid loadedMvid = pathsToAssemblies[path].ManifestModule.ModuleVersionId;
+
+                if (fileMvid != loadedMvid)
+                {
+                    errors.Add(
+                        new ConsistencyError(
+                            ConsistencyErrorKind.LoadedAssemblyDiffers,
+                            path,
+                            pathsToAssemblies[path].FullName));
+                }
+
+                ImmutableArray<AssemblyIdentity> references = pathsToReferences[path];
+                foreach (var reference in references)
+                {
+                    if (!identitiesOfAssembliesOnDisk.Contains(reference) &&
+                        !_prefixWhiteList.Any(prefix => reference.Name.StartsWith(prefix)))
```

Comment:
So we won't complain if there is an inconsistency in the version of System.* / MS.CA.*? Shouldn't we given that there may be behavior changes between versions of these dlls (although hopefully rare)?